worlds/sc2/mission_order/structs.py

```diff
@@ -0,0 +1,932 @@
+        self.option_empty = data.get("empty", self.option_empty)
+        self.option_next = data.get("next", self.option_next)
+        self.option_entry_rules = data.get("entry_rules", self.option_entry_rules)
+        self.option_difficulty = data.get("difficulty", self.option_difficulty)
+        self.option_mission_pool = data.get("mission_pool", self.option_mission_pool)
+    
+    def set_mission(
+            self, world: World, mission: SC2Mission,
+            locations_per_region: Dict[str, List['LocationData']], location_cache: List[Location]
+    ):
+        self.mission = mission
+        self.region = create_region(world, locations_per_region, location_cache,
+                                    mission.mission_name)
+
+    def is_always_unlocked(self) -> bool:
+        return self.entry_rule.is_always_fulfilled()
+    
+    def is_unlocked(self, beaten_missions: Set[SC2MOGenMission]) -> bool:
+        return self.entry_rule.is_fulfilled(beaten_missions)
+    
+    def beat_item(self) -> str:
+        return f"Beat {self.mission.mission_name}"
+    
+    def beat_rule(self, player) -> Callable[[CollectionState], bool]:
+        return lambda state: state.has(self.beat_item(), player)
+
+    def search(self, term: str) -> Union[List[MissionOrderNode], None]:
+        return None
+    
+    def child_type_name(self) -> str:
+        return ""
+
+    def get_missions(self) -> List[SC2MOGenMission]:
+        return [self]
+
+    def get_exits(self) -> List[SC2MOGenMission]:
+        return [self]
+    
+    def get_visual_requirement(self) -> Union[str, SC2MOGenMission]:
+        return self
+    
+    def make_connections(self, world: World, used_names: Dict[str, int]):
+        player = world.player
+        self_rule = self.entry_rule.to_lambda(player)
+        # Only layout entrances need to consider campaign & layout prerequisites
+        if self.option_entrance:
+            campaign_rule = self.parent().parent().entry_rule.to_lambda(player)
+            layout_rule = self.parent().entry_rule.to_lambda(player)
+            unlock_rule = lambda state: campaign_rule(state) and layout_rule(state) and self_rule(state)
+        else:
+            unlock_rule = self_rule
+        # Individually connect to previous missions
+        for mission in self.prev:
+            connect(world, used_names, mission.mission.mission_name, self.mission.mission_name,
+                    lambda state, unlock_rule=unlock_rule: unlock_rule(state))
+        # If there are no previous missions, connect to Menu instead
+        if len(self.prev) == 0:
+            connect(world, used_names, "Menu", self.mission.mission_name,
+                    lambda state, unlock_rule=unlock_rule: unlock_rule(state))
+    
+    def get_slot_data(self) -> MissionSlotData:
+        return MissionSlotData(
+            self.mission.id,
+            {mission.mission.id for mission in self.prev},
+            self.entry_rule.to_slot_data()
```

Comment:
This is because `to_slot_data` is a member of the `EntryRule` ABC, so it can't return a type other than the `RuleData` ABC. `self.entry_rule` on mission order objects will always be a `SubRuleEntryRule`, but I don't know what the best way to handle this situation is.

I can either add a level of abstraction via a method specific to `SubRuleEntryRule` that returns `SubRuleRuleData`, but maybe there's a simpler solution? Maybe something with `TypeVar`s?